docs/design/blocks.md

```diff
@@ -0,0 +1,645 @@
+	// instructing them on what upgrade steps to take.
+	DeprecationMessage string
+}
+```
+
+An example schema similar to blocks (but implemented with nested attributes) would be:
+
+```go
+func (t exampleThingResourceType) GetSchema(_ context.Context) (Schema, diag.Diagnostics) {
+	return Schema{
+		Attributes: map[string]Attribute{
+			"example_attribute": {
+				Optional: true,
+				Computed: true,
+				Attributes: ListNestedAttributes(map[string]Attribute{
+					"example_nested_attribute": {
+						Required: true,
+						Type:     types.StringType,
+					},
+				}, ListNestedAttributesOptions{}),
+			},
+		},
+	}, nil
+}
+```
+
+Which allowed the following configuration:
+
+```terraform
+resource "example_thing" "example" {
+    example_attribute = [
+        {
+            example_nested_attribute = "inside the attribute"
+        }
+    ]
+}
+```
+
+A consideration for block support should be whether the schema definition and value handling can be similar to this existing support to reduce provider developer burden.
+
+## Goals
+
+Block support in this framework should be:
+
+- Available as exported functionality for provider developers.
+- Abstracted from terraform-plugin-go and convertable into those types to separate implementation concerns.
+- Ergonomic to implement Go code (e.g. have helper methods for common use cases).
+
+Additional consideration should be given to:
+
+- Allowing similar value retrieval and writing as attributes, which can enable an easier transition from blocks to nested attributes.
+- Allowing similar schema definitions as attributes, which can enable an easier transition from blocks to nested attributes.
+- Validating any constraints imposed by Terraform, such as lack of sensitivity.
+
+## Proposals
+
+### Blocks as New Nested Attribute Types
+
+Blocks can be implemented within the existing `Attributes` handling inside `Attribute`.
+
+For example:
+
+```go
+"example_block": {
+	Attributes: ListNestedBlocks(map[string]Attribute{
+		"example_attribute": {
+			Type:     types.StringType,
+			Required: true,
+		},
+	}, ListNestedBlocksOptions{}),
+},
+```
+
+While this prevents the addition of a new field on `Attribute`, it can potentially conflate the two concepts and their supported functionality. Future enhancements that only apply to nested attributes may bleed into the overloaded schema definition. Any functionality within the framework would also need to differentiate the `Attributes` types by querying the nesting mode returned by the helper functions, rather than field existance.
+
+### Blocks as New Attribute Field
+
+Blocks can be implemented as a new field inside `Attribute`.
+
+For example:
+
+```go
+"example_block": {
+	Blocks: ListNestedBlocks(map[string]Attribute{
+		"example_attribute": {
+			Type:     types.StringType,
+			Required: true,
+		},
+	}, ListNestedBlocksOptions{}),
+},
+```
+
+This allows the framework to differentiate between block and nested attributes by fields.
```

Comment:
My question here is that I think this could allow providers to define incorrect schemas that we can't model with the protocol. For example, a ListNestedAttributes could contain a `tfsdk.Attribute` that has a `Blocks` property on it, which, if memory serves, we can't actually model in the protocol; attributes can only have attributes under them.

I think allowing this in the type system when it's not supported in the protocol may be confusing.